BWIndex.py

```diff
@@ -0,0 +1,239 @@
+				suffixArray.append(self.SS[nodeIx])
+		self.BWT = transform
+		self.suffixArray = suffixArray
+
+	def drawGraph(self, maxLabelLength=None, outFile=None):
+		D = Digraph()
+		self.NS[0] = 0
+		self.NE[0] = -1
+		for i in range(self.nodeCount):
+			label = self.Text[self.NS[i]:self.NE[i] + 1]
+			if maxLabelLength:
+				label = label[:maxLabelLength]
+			D.node(str(i), label)
+		for tail in self.G:
+			for head in self.G[tail]:
+				D.edge(str(tail), str(head))
+		D.render(outFile, view=True)
+
+	def report(self):
+		print(self.G, self.H, self.NS, self.NE, self.SS, self.BWT, self.suffixArray, self.FirstOccurrence, self.Count)
+
+	def appendNode(self, parent, nodeStringStart, suffixStart):
+		newNode = self.nodeCount
+		self.G[parent] = self.G.get(parent, []) + [newNode]
+		self.H[newNode] = parent
+		self.NS[newNode] = nodeStringStart
+		self.NE[newNode] = len(self.Text) - 1
+		self.SS[newNode] = suffixStart
+		self.nodeCount += 1
+
+	def splitNode(self, node, index):
+		# introduce a new node whose parent is the original parent of node, make it the new parent of node
+		newNode = self.nodeCount
+		parent = self.H[node]
+		self.G[newNode] = [node]
+		self.G[parent].remove(node)
+		self.G[parent] = self.G.get(parent, []) + [newNode]
+		self.H[newNode] = parent
+		self.H[node] = newNode
+		self.NS[newNode] = self.NS[node]
+		self.NE[newNode] = index - 1
+		self.NS[node] = index
+		self.nodeCount += 1
+		return newNode
+
+	def findMatchingChild(self, node, loc):
+		for child in self.G[node]:
+			if self.Text[loc] == self.Text[self.NS[child]]:
+				return child
+		return -1
+
+	def firstMismatch(self, node, ix):
+		offset = 1
+		while True:
+			if ix + offset >= len(self.Text):
+				return self.NS[node] + offset
+			if self.NS[node] + offset > self.NE[node]:
+				return -1
+			if self.Text[self.NS[node] + offset] != self.Text[ix + offset]:
+				return self.NS[node] + offset
+			offset += 1
+
+	def threadSuffix(self, ix):
+		'''
+		look for a node that matches the start of the current substring
+		if none found, create a new node and set its string to the current substring and SS to the initial index
+		if one is found, look for a mismatch within the string
+		if a mismatch is found, split the target node and append the remainder of the substring
+		if no mismatch, advance curIx by the length of the node string and iterate
+		'''
+		node = 0
+		curIx = ix
+		while True:
+			nextNode = self.findMatchingChild(node, curIx)
+			if nextNode == -1:
+				self.appendNode(node, curIx, ix)
+				break
+			node = nextNode
+			splitLoc = self.firstMismatch(node, curIx)
+			if splitLoc == -1:
+				curIx += self.NE[node] - self.NS[node] + 1
+				continue
+			newNode = self.splitNode(node, splitLoc)
+			self.appendNode(newNode, splitLoc + (curIx - self.NS[newNode]), ix)
+			c = self.nodeCount - 1
+			break
+
+	def computeCountArray(self):
+		for i in self.alphabet:
+			self.Count[i] = [0] * (len(self.Text) + 1)  # check does this cause any reference issues?

```

Comment:
Integers are immutable and the spots in the list that are created are separate, so no reference issues here that I'm aware of :)